cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/PushAppsOperation.java

```diff
@@ -0,0 +1,125 @@
+import io.cdap.cdap.internal.operation.LongRunningOperation;
+import io.cdap.cdap.internal.operation.LongRunningOperationContext;
+import io.cdap.cdap.internal.operation.OperationException;
+import io.cdap.cdap.proto.app.UpdateMultiSourceControlMetaReqeust;
+import io.cdap.cdap.proto.app.UpdateSourceControlMetaRequest;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.operation.OperationResource;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.sourcecontrol.ApplicationManager;
+import io.cdap.cdap.sourcecontrol.NoChangesToPushException;
+import io.cdap.cdap.sourcecontrol.SourceControlException;
+import io.cdap.cdap.sourcecontrol.operationrunner.InMemorySourceControlOperationRunner;
+import io.cdap.cdap.sourcecontrol.operationrunner.MultiPushAppOperationRequest;
+import io.cdap.cdap.sourcecontrol.operationrunner.PushAppResponse;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Defines operation for doing SCM Push for connected repositories.
+ **/
+public class PushAppsOperation implements LongRunningOperation {
+
+  private final PushAppsRequest request;
+
+  private final InMemorySourceControlOperationRunner scmOpRunner;
+  private final ApplicationManager applicationManager;
+
+  /**
+   * Only request is passed using AssistedInject. See {@link PushAppsOperationFactory}
+   *
+   * @param request contains apps to push
+   * @param runner runs git operations. The reason we do not use
+   *     {@link io.cdap.cdap.sourcecontrol.operationrunner.SourceControlOperationRunner} rather than
+   *     concrete implementation is because the git operations should always run inMemory.
+   * @param applicationManager provides utilities to provide app-fabric exposed
+   *     functionalities.
+   */
+  @Inject
+  PushAppsOperation(@Assisted PushAppsRequest request,
+      InMemorySourceControlOperationRunner runner,
+      ApplicationManager applicationManager) {
+    this.request = request;
+    this.applicationManager = applicationManager;
+    this.scmOpRunner = runner;
+  }
+
+  @Override
+  public ListenableFuture<Set<OperationResource>> run(LongRunningOperationContext context)
+      throws OperationException {
+    RepositoryConfig repositoryConfig = request.getConfig();
+    NamespaceId namespaceId = context.getRunId().getNamespaceId();
+    MultiPushAppOperationRequest pushReq = new MultiPushAppOperationRequest(
+        namespaceId,
+        repositoryConfig,
+        request.getApps(),
+        request.getCommitDetails()
+    );
+
+    List<PushAppResponse> responses = new ArrayList<>();
+
+    try {
+      responses = scmOpRunner.multiPush(pushReq, applicationManager);
+      context.updateOperationResources(getResources(namespaceId, responses));
+    } catch (SourceControlException | NoChangesToPushException e) {
+      throw new OperationException(
```

Comment:
should pass e as the cause. When wrapping an exception we will basically always want to set the cause if there is one.

Reply:
done